### PR TITLE
doc: Build doxygen documentation

### PR DESCRIPTION
**Pull request by LukasWoodtli**
The Doxygen documentation is generated and uploaded with GitHub actions.

Currently the code is not yet thoroughly documented.

### CODE REVIEW

File: doc/doxygen/Doxyfile [hunk 1]
```diff
@@ -0,0 +1,2673 @@
...
+# are not case sensitive the option should be be set to NO to properly deal with
+# output files written for symbols that only differ in casing, such as for two
+# classes, one named CLASS and the other named Class, and to also support
+# references to files without having to specify the exact matching casing. On
+# Windows (including Cygwin) and MacOS, users should typically set this option
+# to NO, whereas on Linux or other Unix flavors it should typically be set to
+# YES.
+# The default value is: system dependent.
+
+CASE_SENSE_NAMES       = NO
```
Comment (rettichschnidi) [inline doc/doxygen/Doxyfile:597]: CI runs on Linux - should this be Yes or at least the default?

File: doc/doxygen/Doxyfile [hunk 2]
```diff
@@ -0,0 +1,2673 @@
...
+# If the HTML_FORMULA_FORMAT option is set to svg, doxygen will use the pdf2svg
+# tool (see https://github.com/dawbarton/pdf2svg) or inkscape (see
+# https://inkscape.org) to generate formulas as SVG images instead of PNGs for
+# the HTML output. These images will generally look nicer at scaled resolutions.
+# Possible values are: png (the default) and svg (looks nicer but requires the
+# pdf2svg or inkscape tool).
+# The default value is: png.
+# This tag requires that the tag GENERATE_HTML is set to YES.
+
+HTML_FORMULA_FORMAT    = png
```
Comment (rettichschnidi) [inline doc/doxygen/Doxyfile:1623]:
```suggestion
HTML_FORMULA_FORMAT    = svg
```

File: doc/doxygen/Doxyfile [hunk 3]
```diff
@@ -0,0 +1,2673 @@
...
+
+DIAFILE_DIRS           =
+
+# When using plantuml, the PLANTUML_JAR_PATH tag should be used to specify the
+# path where java can find the plantuml.jar file. If left blank, it is assumed
+# PlantUML is not used or called during a preprocessing step. Doxygen will
+# generate a warning when it encounters a \startuml command in this case and
+# will not generate output for the diagram.
+
+PLANTUML_JAR_PATH      =
```
Comment (rettichschnidi) [inline doc/doxygen/Doxyfile:2601]: Would install platnuml via apt and enable support for it:
```suggestion
PLANTUML_JAR_PATH      = /usr/share/plantuml/plantuml.jar
```

File: tools/ci/run_ci.sh [hunk 4]
```diff
@@ -221,6 +223,11 @@ function run_tests() {
   find "${REPO_ROOT_DIR}" -name \*.gcov -exec rm {} \;
 }
 
+function run_doxygen() {
+  mkdir -p build-wakaama/doxygen_out
+    GIT_REVISION=$(git rev-parse @) WORKING_DIR=$(pwd) DOXYGEN_OUT_DIR=build-wakaama/doxygen_out \
```
Comment (rettichschnidi) [inline tools/ci/run_ci.sh:228]: Would use either a dash, such as in build*-*wakaama:
```suggestion
    GIT_REVISION=$(git rev-parse @) WORKING_DIR=$(pwd) DOXYGEN_OUT_DIR=build-wakaama/doxygen-out \
```

Or just leave out the `_out` part:
```suggestion
    GIT_REVISION=$(git rev-parse @) WORKING_DIR=$(pwd) DOXYGEN_OUT_DIR=build-wakaama/doxygen \
```

File: .github/workflows/documentation.yaml [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+        sudo apt-get install doxygen graphviz
+
+    - name: Build Doxygen documentation
+      run: tools/ci/run_ci.sh --run-doxygen
+
+    - name: Upload Doxygen documentation
+      uses: actions/upload-artifact@v4
+      with:
+        name: Doxygen documentation (HTML)
+        path: build-wakaama/doxygen_out
```
Comment (rettichschnidi) [inline .github/workflows/documentation.yaml:30]:
```suggestion
        path: build-wakaama/doxygen-out
```

or
```suggestion
        path: build-wakaama/doxygen
```

(see other comment)

**Review by rettichschnidi — CHANGES_REQUESTED** [review]

File: .github/workflows/documentation.yaml [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+jobs:
+  coverage:
+    runs-on: ubuntu-24.04
+
+    steps:
+    - name: Checkout code including full history and submodules
+      uses: actions/checkout@v4
+      with:
+        submodules: true
```
Comment (rettichschnidi) [inline .github/workflows/documentation.yaml:15]: I just checked out without submodules and build the documentation. It does not seem like submodules are actually needed.

File: doc/doxygen/Doxyfile
Comment (rettichschnidi) [inline doc/doxygen/Doxyfile:1]: Looking really good.

One tiny nit, likely not worth investing time into: The tab "Related Pages" shows the exact same content as "Main Page".

image

**Review by rettichschnidi — APPROVED** [review]
LGTM, just two nits.

File: doc/doxygen/Doxyfile
Comment (LukasWoodtli) [reply doc/doxygen/Doxyfile:1]: Now the "Related Pages" tab shows additional files:
image

Comment (rettichschnidi) [reply doc/doxygen/Doxyfile:1]: All links are still on the same page (README.md), but this clearly an improvement.

**Pull request merged**